File: src/IECore/ImagePrimitive.cpp
//////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2007-2013, Image Engine Design Inc. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//
//     * Neither the name of Image Engine Design nor the names of any
//       other contributors to this software may be used to endorse or
//       promote products derived from this software without specific prior
//       written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//////////////////////////////////////////////////////////////////////////

#include <cassert>

#include "boost/static_assert.hpp"

#include "IECore/ImagePrimitive.h"
#include "IECore/MessageHandler.h"
#include "IECore/Renderer.h"
#include "IECore/TypeTraits.h"
#include "IECore/DespatchTypedData.h"
#include "IECore/MurmurHash.h"

using namespace std;
using namespace IECore;
using namespace Imath;
using namespace boost;

static IndexedIO::EntryID g_displayWindowMinXEntry("displayWindowMinX");
static IndexedIO::EntryID g_displayWindowMinYEntry("displayWindowMinY");
static IndexedIO::EntryID g_displayWindowMaxXEntry("displayWindowMaxX");
static IndexedIO::EntryID g_displayWindowMaxYEntry("displayWindowMaxY");
static IndexedIO::EntryID g_dataWindowMinXEntry("dataWindowMinX");
static IndexedIO::EntryID g_dataWindowMinYEntry("dataWindowMinY");
static IndexedIO::EntryID g_dataWindowMaxXEntry("dataWindowMaxX");
static IndexedIO::EntryID g_dataWindowMaxYEntry("dataWindowMaxY");

const unsigned int ImagePrimitive::m_ioVersion = 1;
IE_CORE_DEFINEOBJECTTYPEDESCRIPTION(ImagePrimitive);

ImagePrimitive::ImagePrimitive()
{
}

ImagePrimitive::ImagePrimitive( const Box2i &dataWindow, const Box2i &displayWindow )
{
	setDataWindow( dataWindow );
	setDisplayWindow( displayWindow );
}

Box3f ImagePrimitive::bound() const
{
	assert( ! m_displayWindow.isEmpty() );

	/// \todo We might need to include any pixel aspect ratio in this bound

	V3f boxMin( m_displayWindow.min.x, m_displayWindow.min.y, 0.0 );

	/// We add one here because the displayWindow is measured in pixels, and is inclusive. That is, an image
	/// which has a displayWindow of (0,0)->(0,0) contains exactly one pixel.

	V3f boxMax( 1.0f + m_displayWindow.max.x, 1.0f + m_displayWindow.max.y, 0.0 );

	V3f center = (boxMin + boxMax) / 2.0;

	return Box3f( boxMin - center, boxMax - center );
}

const Box2i &ImagePrimitive::getDataWindow() const
{
	return m_dataWindow;
}

void ImagePrimitive::setDataWindow( const Box2i &dataWindow )
{
	m_dataWindow = dataWindow;
}

const Box2i &ImagePrimitive::getDisplayWindow() const
{
	return m_displayWindow;
}

void ImagePrimitive::setDisplayWindow( const Box2i &displayWindow )
{
	if ( displayWindow.isEmpty() )
	{
		throw InvalidArgumentException( "ImagePrimitive: Cannot set displayWindow to the empty window" );
	}

	m_displayWindow = displayWindow;
}

// give the size of the image
size_t ImagePrimitive::variableSize( PrimitiveVariable::Interpolation interpolation ) const
{
	switch (interpolation)
	{

	case PrimitiveVariable::Vertex:
	case PrimitiveVariable::Varying:
	case PrimitiveVariable::FaceVarying:
		return ( 1 + m_dataWindow.max.x - m_dataWindow.min.x ) * ( 1 + m_dataWindow.max.y - m_dataWindow.min.y );

	default:
		return 1;

	}
}

void ImagePrimitive::render( Renderer *renderer ) const
{
	assert( renderer );

	renderer->image(m_dataWindow, m_displayWindow, variables);
}

//
// handling for serialization
//
void ImagePrimitive::copyFrom( const IECore::Object *rhs, IECore::Object::CopyContext *context )
{
	assert( rhs );
	assert( context );

	Primitive::copyFrom(rhs, context);
	const ImagePrimitive *p_rhs = static_cast<const ImagePrimitive *>(rhs);

	m_displayWindow = p_rhs->getDisplayWindow();
	m_dataWindow = p_rhs->getDataWindow();
}

void ImagePrimitive::save(IECore::Object::SaveContext *context) const
{
	assert( context );

	Primitive::save(context);
	IndexedIOPtr container = context->container(staticTypeName(), m_ioVersion);

	container->write(g_displayWindowMinXEntry, m_displayWindow.min.x);
	container->write(g_displayWindowMinYEntry, m_displayWindow.min.y);
	container->write(g_displayWindowMaxXEntry, m_displayWindow.max.x);
	container->write(g_displayWindowMaxYEntry, m_displayWindow.max.y);

	container->write(g_dataWindowMinXEntry, m_dataWindow.min.x);
	container->write(g_dataWindowMinYEntry, m_dataWindow.min.y);
	container->write(g_dataWindowMaxXEntry, m_dataWindow.max.x);
	container->write(g_dataWindowMaxYEntry, m_dataWindow.max.y);
}

void ImagePrimitive::load(IECore::Object::LoadContextPtr context)
{
	assert( context );

	Primitive::load(context);
	unsigned int v = m_ioVersion;

	ConstIndexedIOPtr container = context->container(staticTypeName(), v);

	container->read(g_displayWindowMinXEntry, m_displayWindow.min.x);
	container->read(g_displayWindowMinYEntry, m_displayWindow.min.y);
	container->read(g_displayWindowMaxXEntry, m_displayWindow.max.x);
	container->read(g_displayWindowMaxYEntry, m_displayWindow.max.y);

	if ( v < 1 )
	{
		m_dataWindow = m_displayWindow;
	}
	else
	{
		container->read(g_dataWindowMinXEntry, m_dataWindow.min.x);
		container->read(g_dataWindowMinYEntry, m_dataWindow.min.y);
		container->read(g_dataWindowMaxXEntry, m_dataWindow.max.x);
		container->read(g_dataWindowMaxYEntry, m_dataWindow.max.y);
	}
}

bool ImagePrimitive::isEqualTo(const IECore::Object *rhs) const
{
	assert( rhs );

	if (!Primitive::isEqualTo(rhs))
	{
		return false;
	}

	const ImagePrimitive *p_rhs = static_cast<const ImagePrimitive *>(rhs);

	// return true iff we have the same data window.
	// this is not complete
	return m_dataWindow == p_rhs->getDataWindow() && m_displayWindow == p_rhs->getDisplayWindow();
}

void ImagePrimitive::memoryUsage( Object::MemoryAccumulator &a ) const
{
	Primitive::memoryUsage( a );
	a.accumulate( sizeof(m_displayWindow) );
	a.accumulate( sizeof(m_dataWindow) );
}

void ImagePrimitive::hash( MurmurHash &h ) const
{
	Primitive::hash( h );
}

void ImagePrimitive::topologyHash( MurmurHash &h ) const
{
	h.append( m_dataWindow );
	h.append( m_displayWindow );
}

///////////////////////////////////////////////////////////////////////////////////////////
// Space methods
///////////////////////////////////////////////////////////////////////////////////////////

M33f ImagePrimitive::objectToUVMatrix() const
{
	M33f result;
	const Box2i &displayWindow = getDisplayWindow();
	V2i size = displayWindow.size() + V2i( 1 );
	result.translate( V2f( 0.5f ) );
	result.scale( V2f( 1.0f ) / V2f( size.x, -size.y ) );
	return result;
}

M33f ImagePrimitive::uvToObjectMatrix() const
{
	M33f result;
	const Box2i &displayWindow = getDisplayWindow();
	V2i size = displayWindow.size() + V2i( 1 );
	result.scale( V2f( size.x, -size.y ) );
	result.translate( V2f( -0.5f ) );
	return result;
}

M33f ImagePrimitive::objectToPixelMatrix() const
{
	M33f result;
	const Box2i &displayWindow = getDisplayWindow();
	V2i size = displayWindow.size();
	result.translate( V2f( displayWindow.min.x, displayWindow.min.y ) + V2f( size.x, size.y ) / 2.0f );
	result.scale( V2f( 1.0f, -1.0f ) );
	return result;
}

M33f ImagePrimitive::pixelToObjectMatrix() const
{
	M33f result;
	const Box2i &displayWindow = getDisplayWindow();
	V2i size = displayWindow.size();
	result.scale( V2f( 1.0f, -1.0f ) );
	result.translate( -V2f( displayWindow.min.x, displayWindow.min.y ) - V2f( size.x, size.y ) / 2.0f );
	return result;
}
		
M33f ImagePrimitive::pixelToUVMatrix() const
{
	M33f result;
	const Box2i &displayWindow = getDisplayWindow();
	V2i size = displayWindow.size() + V2i( 1 );
	result.scale( V2f( 1.0f ) / V2f( size.x, size.y ) );
	result.translate( V2f( 0.5f ) - V2f( displayWindow.min.x, displayWindow.min.y ) );
	return result;
}

M33f ImagePrimitive::uvToPixelMatrix() const
{
	M33f result;
	const Box2i &displayWindow = getDisplayWindow();
	V2i size = displayWindow.size() + V2i( 1 );
	result.translate( V2f( displayWindow.min.x, displayWindow.min.y ) - V2f( 0.5f ) );
	result.scale( V2f( size.x, size.y ) );
	return result;
}

Imath::M33f ImagePrimitive::matrix( Space inputSpace, Space outputSpace ) const
{
	switch( inputSpace )
	{
		case Pixel :
			
			switch( outputSpace )
			{
				case Pixel :
					return M33f();
				case UV :
					return pixelToUVMatrix();
				case Object :
					return pixelToObjectMatrix();
				default :
					throw Exception( "Unknown output space" );
			}
			
		case UV :
		
			switch( outputSpace )
			{
				case Pixel :
					return uvToPixelMatrix();
				case UV :
					return M33f();
				case Object :
					return uvToObjectMatrix();
				default :
					throw Exception( "Unknown output space" );
			}
		
		case Object :
	
			switch( outputSpace )
			{
				case Pixel :
					return objectToPixelMatrix();
				case UV :
					return objectToUVMatrix();
				case Object :
					return M33f();
				default :
					throw Exception( "Unknown output space" );
			}
		
		default :
			
			throw Exception( "Unknown input space" );
			
	};
}
		
///////////////////////////////////////////////////////////////////////////////////////////
// Channel methods
///////////////////////////////////////////////////////////////////////////////////////////

bool ImagePrimitive::channelValid( const PrimitiveVariable &pv, std::string *reason ) const
{
	if( pv.interpolation!=PrimitiveVariable::Vertex &&
		pv.interpolation!=PrimitiveVariable::Varying &&
		pv.interpolation!=PrimitiveVariable::FaceVarying )
	{
		if( reason )
		{
			*reason = "Primitive variable has inappropriate interpolation.";
		}
		return false;
	}
	if( !pv.data )
	{
		if( reason )
		{
			*reason = "Primitive variable has no data.";
		}
		return false;
	}

	if( !despatchTraitsTest<TypeTraits::IsNumericVectorTypedData>( pv.data.get() ) )
	{
		if( reason )
		{
			*reason = "Primitive variable has inappropriate type.";
		}
		return false;
	}

	size_t size = despatchTypedData<TypedDataSize>( pv.data.get() );
	size_t numPixels = variableSize( PrimitiveVariable::Vertex );
	if( size!=numPixels )
	{
		if( reason )
		{
			*reason = str( format( "Primitive variable has wrong size (%d but should be %d)." ) % size % numPixels );
		}
		return false;
	}

	return true;
}

bool ImagePrimitive::channelValid( const std::string &name, std::string *reason ) const
{
	PrimitiveVariableMap::const_iterator it = variables.find( name );
	if( it==variables.end() )
	{
		if( reason )
		{
			*reason = str( format( "Primitive variable \"%s\" does not exist." ) % name );
		}
		return false;
	}
	return channelValid( it->second, reason );
}

void ImagePrimitive::channelNames( vector<string> &names ) const
{
	// copy in the names of channels from the map
	names.clear();

	for ( PrimitiveVariableMap::const_iterator i = variables.begin(); i != variables.end(); ++i )
	{
		if( channelValid( i->second ) )
		{
			names.push_back( i->first );
		}
	}
}

File: test/IECore/ImagePrimitive.py
##########################################################################
#
#  Copyright (c) 2007-2013, Image Engine Design Inc. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#
#     * Neither the name of Image Engine Design nor the names of any
#       other contributors to this software may be used to endorse or
#       promote products derived from this software without specific prior
#       written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################

import unittest
import os

from IECore import *

class TestImagePrimitive( unittest.TestCase ) :

	def testConstructor( self ) :
		""" Test ImagePrimitive constructor """

		windowMin = V2i( 0, 0 )
		windowMax = V2i( 100, 100 )
		w = Box2i( windowMin, windowMax )
		i = ImagePrimitive( w, w )

		self.assertEqual( i.dataWindow, w )
		self.assertEqual( i.displayWindow, w )

		i.dataWindow = Box2i( windowMin, V2i( 10, 10 ) )
		self.assertEqual( i.dataWindow, Box2i( windowMin, V2i( 10, 10 ) ) )
		self.assertEqual( i.displayWindow, w )

		i.displayWindow = Box2i( windowMin, V2i( 10, 10 ) )
		self.assertEqual( i.displayWindow, Box2i( windowMin, V2i( 10, 10 ) ) )

	def testDataWindow( self ) :

		displayWindow = Box2i( V2i( 0, 0 ), V2i( 99, 99 ) )
		dataWindow = Box2i( V2i( 50, 50), V2i( 99, 99 ) )
		img = ImagePrimitive( dataWindow, displayWindow )

	def testBound( self ) :
		""" Test ImagePrimitive bound """

		windowMin = V2i( 0, 0 )
		windowMax = V2i( 99, 99 )
		w = Box2i( windowMin, windowMax )
		i = ImagePrimitive( w, w )

		self.assertEqual( i.bound(), Box3f( V3f( -50, -50, 0 ), V3f( 50, 50, 0 ) ) )

		windowMin = V2i( 50, 50 )
		windowMax = V2i( 99, 99 )
		w = Box2i( windowMin, windowMax )
		i = ImagePrimitive( w, w )

		self.assertEqual( i.bound(), Box3f( V3f( -25, -25, 0 ), V3f( 25, 25, 0 ) ) )

	def testDataWindow( self ) :
		""" Test ImagePrimitive data window """

		displayWindow = Box2i( V2i( 0, 0 ), V2i( 99, 99 ) )
		dataWindow = Box2i( V2i( 50, 50), V2i( 99, 99 ) )
		img = ImagePrimitive( dataWindow, displayWindow )

		dataWindowArea = 50 * 50
		R = FloatVectorData( dataWindowArea )
		G = FloatVectorData( dataWindowArea )
		B = FloatVectorData( dataWindowArea )

		img["R"] = PrimitiveVariable( PrimitiveVariable.Interpolation.Vertex, R )
		img["G"] = PrimitiveVariable( PrimitiveVariable.Interpolation.Vertex, G )
		img["B"] = PrimitiveVariable( PrimitiveVariable.Interpolation.Vertex, B )

		self.assert_( img.arePrimitiveVariablesValid() )

		# \todo Verify behaviour when dataWindow and displayWindow are contradictory or inconsistent

	def testLoadSave( self ):
		""" Test ImagePrimitive load/save """

		windowMin = V2i( 0, 0 )
		windowMax = V2i( 100, 100 )
		w = Box2i( windowMin, windowMax )
		i = ImagePrimitive( w, w )

		Writer.create( i, "test/IECore/data/output.cob" ).write()

		i2 = Reader.create( "test/IECore/data/output.cob" ).read()
		self.assertEqual( type(i2), ImagePrimitive )

		self.assertEqual( i.displayWindow, i2.displayWindow )
		self.assertEqual( i.dataWindow, i2.dataWindow )

	def testChannelNames( self ) :

		""" Test ImagePrimitive channel names """

		windowMin = V2i( 0, 0 )
		windowMax = V2i( 99, 99 )
		w = Box2i( windowMin, windowMax )
		i = ImagePrimitive( w, w )

		r = FloatVectorData()
		r.resize( 100 * 100 )

		i["R"] = PrimitiveVariable( PrimitiveVariable.Interpolation.Vertex, r )

		self.assert_( "R" in i.channelNames() )

		b = FloatData()

		i["B"] = PrimitiveVariable( PrimitiveVariable.Interpolation.Constant, b )

		self.failIf( "B" in i.channelNames() )

		self.assert_( i.arePrimitiveVariablesValid() )

		# Deliberately make a primvar too small!
		g = FloatVectorData()
		g.resize( 50 * 100 )

		i["G"] = PrimitiveVariable( PrimitiveVariable.Interpolation.Vertex, g )

		self.failIf( "G" in i.channelNames() )
		self.failIf( i.arePrimitiveVariablesValid() )

	def testCreateChannel( self ):

		windowMin = V2i( 0, 0 )
		windowMax = V2i( 99, 99 )
		w = Box2i( windowMin, windowMax )
		i = ImagePrimitive( w, w )

		i.createFloatChannel( "R" )
		i.createHalfChannel( "G" )
		i.createUIntChannel( "B" )

		self.assert_( "R" in i )
		self.assert_( "G" in i )
		self.assert_( "B" in i )

	def testErrors( self ):

		windowMin = V2i( 0, 0 )
		windowMax = V2i( 99, 99 )
		w = Box2i( windowMin, windowMax )
		i = ImagePrimitive( w, w )

		empty = Box2i()

		self.assertRaises( RuntimeError, setattr, i, "displayWindow", empty )

		self.assertRaises( RuntimeError, ImagePrimitive, empty, empty )

	def testChannelValid( self ) :

		b = Box2i( V2i( 0 ), V2i( 9 ) )
		i = ImagePrimitive( b, b )

		d = FloatVectorData( [ 1 ] )

		p = PrimitiveVariable( PrimitiveVariable.Interpolation.Uniform, d )
		i["Y"] = p

		self.assertEqual( i.channelValid( p ), False )
		self.assertEqual( i.channelValid( "Y" ), False )
		self.assertEqual( i.getChannel( "Y" ), None )

		t = i.channelValid( p, True )
		self.assert_( isinstance( t, tuple ) )
		self.assertEqual( t[0], False )
		self.assert_( isinstance( t[1], str ) )

		p = PrimitiveVariable( PrimitiveVariable.Interpolation.Constant, d )
		i["Y"] = p

		self.assertEqual( i.channelValid( p ), False )
		self.assertEqual( i.channelValid( "Y" ), False )
		self.assertEqual( i.getChannel( "Y" ), None )

		p = PrimitiveVariable( PrimitiveVariable.Interpolation.Vertex, d )
		i["Y"] = p

		self.assertEqual( i.channelValid( p ), False )
		self.assertEqual( i.channelValid( "Y" ), False )
		self.assertEqual( i.getChannel( "Y" ), None )

		d.resize( 100 )

		self.assertEqual( i.channelValid( p ), True )
		self.assertEqual( i.channelValid( "Y" ), True )
		self.assert_( d.isSame( i.getChannel( "Y" ) ) )

		pp = PrimitiveVariable( PrimitiveVariable.Interpolation.Vertex, FloatData( 1 ) )
		self.assertEqual( i.channelValid( pp ), False )
		i["PP"] = pp
		self.assertEqual( i.channelValid( "PP" ), False )
		self.assertEqual( i.getChannel( "PP" ), None )
	
	def testConvenienceConstructors( self ) :
	
		""" Test ImagePrimitive convenience constructors """
		
		window1Min = V2i(  0,  0 )
		window1Max = V2i( 15, 15 )		
		w1 = Box2i( window1Min, window1Max )
		
		window2Min = V2i(  4,  4 )
		window2Max = V2i( 11, 11 )		
		w2 = Box2i( window2Min, window2Max )
		
		fill = Color3f( 0.49, 0.50, 0.51 )
		i = ImagePrimitive.createRGBFloat( fill, w1, w2 )

		self.assert_( i.isInstanceOf( ImagePrimitive.staticTypeId() ) )
	
		self.assert_( "R" in i )
		self.assert_( "G" in i )
		self.assert_( "B" in i )
		self.assert_( "Y" not in i )

		self.assertEqual( i.dataWindow, w1 )
		self.assertEqual( i.displayWindow, w2 )
		
		self.assert_( i["R"].data.isInstanceOf( FloatVectorData.staticTypeId() ) )
		self.assert_( i["G"].data.isInstanceOf( FloatVectorData.staticTypeId() ) )
		self.assert_( i["B"].data.isInstanceOf( FloatVectorData.staticTypeId() ) )

		self.assertEqual( i["R"].data.size(), 256 )	
		self.assertEqual( i["G"].data.size(), 256 )	
		self.assertEqual( i["B"].data.size(), 256 )	

		for p in ( 0, 63, 127, 255 ) :
			self.assertEqual( i["R"].data[ p ], fill[0] )
			self.assertEqual( i["G"].data[ p ], fill[1] )
			self.assertEqual( i["B"].data[ p ], fill[2] )
		
		fill = 0.5	
		i = ImagePrimitive.createGreyscaleFloat( fill, w1, w2 )
		
		self.assert_( i.isInstanceOf( ImagePrimitive.staticTypeId() ) )

		self.assert_( "R" not in i )
		self.assert_( "G" not in i )
		self.assert_( "B" not in i )
		self.assert_( "Y" in i )

		self.assertEqual( i.dataWindow, w1 )
		self.assertEqual( i.displayWindow, w2 )
		
		self.assert_( i["Y"].data.isInstanceOf( FloatVectorData.staticTypeId() ) )
		self.assertEqual( i["Y"].data.size(), 256 )	

		for p in ( 0, 63, 127, 255 ) :
			self.assertEqual( i["Y"].data[ p ], fill )
	
	def testSpaces( self ) :
	
		# one pixel image 0,0 -> 0,0
	
		onePixelWindow = Box2i( V2i( 0 ), V2i( 0 ) )
		i = ImagePrimitive( onePixelWindow, onePixelWindow )
		
		m = i.pixelToObjectMatrix()
		self.assertEqual( V2f( 0 ) * m, V2f( 0 ) )
		m2 = i.objectToPixelMatrix()
		self.assertEqual( m2, m.inverse() )
		
		m = i.pixelToUVMatrix()
		self.assertEqual( V2f( 0 ) * m, V2f( 0.5 ) )
		m2 = i.uvToPixelMatrix()
		self.assertEqual( m2, m.inverse() )
		
		m = i.objectToUVMatrix()
		self.assertEqual( V2f( -0.5 ) * m, V2f( 0, 1 ) )
		self.assertEqual( V2f( 0.5 ) * m, V2f( 1, 0 ) )
		m2 = i.uvToObjectMatrix()
		self.assertEqual( m2, m.inverse() )
		
		self.failUnless( ( i.objectToUVMatrix() * i.uvToPixelMatrix() ).equalWithAbsError( i.objectToPixelMatrix(), 0.00001 ) )
		self.failUnless( ( i.pixelToUVMatrix() * i.uvToObjectMatrix() ).equalWithAbsError( i.pixelToObjectMatrix(), 0.00001 ) )
		
		self.assertEqual( i.matrix( ImagePrimitive.Space.UV, ImagePrimitive.Space.Pixel ), i.uvToPixelMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.UV, ImagePrimitive.Space.Object ), i.uvToObjectMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Pixel, ImagePrimitive.Space.UV ), i.pixelToUVMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Pixel, ImagePrimitive.Space.Object ), i.pixelToObjectMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Object, ImagePrimitive.Space.Pixel ), i.objectToPixelMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Object, ImagePrimitive.Space.UV ), i.objectToUVMatrix() )
				
		# two pixel image 0,0 -> 1,1

		twoPixelWindow = Box2i( V2i( 0 ), V2i( 1 ) )
		i = ImagePrimitive( twoPixelWindow, twoPixelWindow )
		
		m = i.pixelToObjectMatrix()
		self.assertEqual( V2f( 0 ) * m, V2f( -0.5, 0.5 ) )
		self.assertEqual( V2f( 1 ) * m, V2f( 0.5, -0.5 ) )
		m2 = i.objectToPixelMatrix()
		self.assertEqual( m2, m.inverse() )
		
		m = i.pixelToUVMatrix()
		self.assertEqual( V2f( 0 ) * m, V2f( 0.25 ) )
		self.assertEqual( V2f( 1 ) * m, V2f( 0.75 ) )
		m2 = i.uvToPixelMatrix()
		self.assertEqual( m2, m.inverse() )
		
		m = i.objectToUVMatrix()
		self.assertEqual( V2f( -1 ) * m, V2f( 0, 1 ) )
		self.assertEqual( V2f( 1 ) * m, V2f( 1, 0 ) )
		m2 = i.uvToObjectMatrix()
		self.assertEqual( m2, m.inverse() )
		
		self.failUnless( ( i.objectToUVMatrix() * i.uvToPixelMatrix() ).equalWithAbsError( i.objectToPixelMatrix(), 0.00001 ) )
		self.failUnless( ( i.pixelToUVMatrix() * i.uvToObjectMatrix() ).equalWithAbsError( i.pixelToObjectMatrix(), 0.00001 ) )

		self.assertEqual( i.matrix( ImagePrimitive.Space.UV, ImagePrimitive.Space.Pixel ), i.uvToPixelMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.UV, ImagePrimitive.Space.Object ), i.uvToObjectMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Pixel, ImagePrimitive.Space.UV ), i.pixelToUVMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Pixel, ImagePrimitive.Space.Object ), i.pixelToObjectMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Object, ImagePrimitive.Space.Pixel ), i.objectToPixelMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Object, ImagePrimitive.Space.UV ), i.objectToUVMatrix() )

		# three by two pixel image 10,20 -> 12,21

		threeTwoPixelWindowOffset = Box2i( V2i( 10, 20 ), V2i( 12, 21 ) )
		i = ImagePrimitive( threeTwoPixelWindowOffset, threeTwoPixelWindowOffset )
		
		m = i.pixelToObjectMatrix()
		self.assertEqual( V2f( 10, 20 ) * m, V2f( -1, 0.5 ) )
		self.assertEqual( V2f( 12, 21 ) * m, V2f( 1, -0.5 ) )
		m2 = i.objectToPixelMatrix()
		self.assertEqual( m2, m.inverse() )
		
		m = i.pixelToUVMatrix()
		self.failUnless( ( V2f( 10, 20 ) * m ).equalWithAbsError( V2f( 1/6.0, 1/4.0 ), 0.00001 ) )
		self.failUnless( ( V2f( 12, 21 ) * m ).equalWithAbsError( V2f( 5/6.0, 3/4.0 ), 0.00001 ) )
		m2 = i.uvToPixelMatrix()
		self.assertEqual( m2, m.inverse() )
		
		m = i.objectToUVMatrix()
		self.assertEqual( V2f( -1.5, -1 ) * m, V2f( 0, 1 ) )
		self.assertEqual( V2f( 1.5, 1 ) * m, V2f( 1, 0 ) )
		m2 = i.uvToObjectMatrix()
		self.assertEqual( m2, m.inverse() )
		
		self.failUnless( ( i.objectToUVMatrix() * i.uvToPixelMatrix() ).equalWithAbsError( i.objectToPixelMatrix(), 0.00001 ) )
		self.failUnless( ( i.pixelToUVMatrix() * i.uvToObjectMatrix() ).equalWithAbsError( i.pixelToObjectMatrix(), 0.00001 ) )

		self.assertEqual( i.matrix( ImagePrimitive.Space.UV, ImagePrimitive.Space.Pixel ), i.uvToPixelMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.UV, ImagePrimitive.Space.Object ), i.uvToObjectMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Pixel, ImagePrimitive.Space.UV ), i.pixelToUVMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Pixel, ImagePrimitive.Space.Object ), i.pixelToObjectMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Object, ImagePrimitive.Space.Pixel ), i.objectToPixelMatrix() )
		self.assertEqual( i.matrix( ImagePrimitive.Space.Object, ImagePrimitive.Space.UV ), i.objectToUVMatrix() )
	
	def testHash( self ) :
	
		w = Box2i( V2i( 0 ), V2i( 10 ) )
		i = ImagePrimitive( w, w )
		h = i.hash()
		t = i.topologyHash()
		
		i.displayWindow = Box2i( V2i( 10 ), V2i( 20 ) )
		self.assertNotEqual( i.hash(), h )
		self.assertNotEqual( i.topologyHash(), h )
		h = i.hash()
		t = i.topologyHash()
		
		i.dataWindow = Box2i( V2i( 10 ), V2i( 20 ) )
		self.assertNotEqual( i.hash(), h )
		self.assertNotEqual( i.topologyHash(), h )
		h = i.hash()
		t = i.topologyHash()
		
		i["primVar"] = PrimitiveVariable( PrimitiveVariable.Interpolation.Constant, IntData( 10 ) )
		self.assertNotEqual( i.hash(), h )
		self.assertEqual( i.topologyHash(), t )
		
	def tearDown( self ) :

		if os.path.exists( "test/IECore/data/output.cob" ) :

			os.remove( "test/IECore/data/output.cob" )


if __name__ == "__main__":
    unittest.main()
